data_structures/sparse_table.cpp

```diff
@@ -0,0 +1,136 @@
+/**
+ * @file
+ * @brief Implementation of [Sparse
+ * Table](https://brilliant.org/wiki/sparse-table/) for `min()` function.
+ * @author [Mann Patel](https://github.com/manncodes)
+ * @details
+ * Sparse Table is a data structure, that allows answering range queries.
+ * It can answer most range queries in O(logn), but its true power is answering
+ * range minimum queries (or equivalent range maximum queries). For those
+ * queries it can compute the answer in O(1) time. The only drawback of this
+ * data structure is, that it can only be used on immutable arrays. This means,
+ * that the array cannot be changed between two queries.
+ *
+ * If any element in the array changes, the complete data structure has to be
+ * recomputed.
+ *
+ * @warning
+ * This sparse table is made for `min(a1,a2,...an)` duplicate invariant
+ * function. This implementation can be changed to other functions like
+ * `gcd()`, `lcm()`, and `max()` by changing a few lines of code.
+ */
+
+#include <array>     /// for std::array
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+
+/**
+ * @namespace data_structures
+ * @brief Data Structures algorithms
+ */
+namespace data_structures {
+
+/**
+ * @namespace sparse_table
+ * @brief sparse table and it's util functions.
+ */
+namespace sparse_table {
+
+/**
+ * @brief A struct to represent sparse table for `min()` as their invariant
+ * function, for the given array `A`. The answer to queries are stored in array
+ * ST.
+ */
+struct Sparse_table {
+    const static int N = 12345;  ///< the maximum size of the array.
+    const static int M = 14;     ///< ceil(log2(N)).
+    int n;                       ///< size of input array.
```

Comment:
It's always best to initialize variables to `0`, as they can display wrong values in some cases.
```suggestion
    int n = 0;                       ///< size of input array.
```

Reply:
should I initialize other arrays to zero to?